File: 06-week-5--big-o-continued/00-days-1-to-2--implement-a-stack-class/solutions/stack.js
class Stack {
  constructor() {
    this.stack = [];
    // this is an arbitrary value to make testing easier
    // 1,024 would be way too high to test!
    this.limit = 10;
  }

  // add item to top of stack if not full
  push(item) {
    if (!this.isFull()) {
      this.stack.push(item);
    } else {
      throw new Error('Stack is full!');
    }
  }

  // remove item from top of stack
  pop() {
    return this.stack.pop();
  }

  // return item at top of stack without removing it
  peek() {
    return this.stack[this.size() - 1];
  }

  // return true if stack is empty, otherwise false
  isEmpty() {
    return this.size() === 0;
  }

  // return true if stack is full, otherwise false
  isFull() {
    return this.size() === this.limit;
  }

  // return number of items in stack
  size() {
    return this.stack.length;
  }

  // return -1 if item not in stack, otherwise integer representing 
  // how far it is from the top
  search(target) {
    for (let i = -1; i >= -this.size(); --i) {
      if (this.stack[this.size() + i] === target) {
        return Math.abs(i) - 1;
      }
    }

    return -1;
  }

  // print contents of stack: do not return the stack itself!
  print() {
    console.log(this.stack.join(' <- '));
  }
}

if (require.main === module) {
  // add your own tests in here
  const stack = new Stack();

  console.log('size', stack.size());
  console.log('is empty?', stack.isEmpty());
  console.log('is full?', stack.isFull());
  console.log('find 5', stack.search(5));
  console.log('peek while empty', stack.peek());

  for (let i = 0; i < 5; ++i) {
    stack.push(i);
  }

  console.log('ADD ITEMS 0 TO 4');
  console.log('size', stack.size());
  console.log('is empty?', stack.isEmpty());
  console.log('is full?', stack.isFull());
  console.log('find 3', stack.search(3));
  console.log('peek', stack.peek());
  stack.print();

  for (let i = 5; i < 10; ++i) {
    stack.push(i);
  }

  console.log('ADD ITEMS 5 TO 9');
  console.log('size', stack.size());
  console.log('is empty?', stack.isEmpty());
  console.log('is full?', stack.isFull());
  console.log('find 3', stack.search(3));
  console.log('peek', stack.peek());
  stack.print();

  console.log('pop', stack.pop());
  console.log('size', stack.size());
  console.log('is empty?', stack.isEmpty());
  console.log('is full?', stack.isFull());
  console.log('peek', stack.peek());
  stack.print();

  console.log('GENERATE ERROR');
  stack.push(9);

  try {
    stack.push(10);
  } catch(err) {
    console.log(err);
    stack.print();
  }
}

module.exports = Stack;

File: 06-week-5--big-o-continued/00-days-1-to-2--implement-a-stack-class/solutions/stack.rb
class Stack
  attr_reader :limit

  def initialize
    @stack = []
    # this is an arbitrary value to make testing easier
    # 1,024 would be way too high to test!
    @limit = 10
  end

  # add item to top of stack if not full
  # if full, throw error
  def push(item)
    raise 'Stack is full' if isFull?

    @stack.push(item)
  end

  # remove item from top of stack and return it
  def pop
    @stack.pop
  end

  # return item at top of stack without removing it
  def peek
    @stack.last
  end

  # return true if stack is empty, otherwise false
  def isEmpty?
    size.zero?
  end

  # return true if stack is full, otherwise false
  def isFull?
    size === @limit
  end

  # return number of items in stack
  def size
    @stack.length
  end

  # return -1 if item not in stack, otherwise integer representing 
  # how far it is from the top
  def search(target)
    @stack.each_with_index do |item, idx|
      return size - idx - 1 if item == target
    end

    -1
  end

  # print contents of stack: do not return the stack itself!
  def print
    puts @stack.join(' <- ')
  end
end

if __FILE__ == $PROGRAM_NAME
  stack = Stack.new;

  puts "'size': #{stack.size}"
  puts "'is empty?': #{stack.isEmpty?}"
  puts "'is full?': #{stack.isFull?}"
  puts "'find 5': #{stack.search(5)}"
  puts "'peek while empty': #{stack.peek}"

  (0...5).each do |num|
    stack.push(num)
  end

  puts "'ADD ITEMS 0 TO 4'"
  puts "'size': #{stack.size}"
  puts "'is empty?': #{stack.isEmpty?}"
  puts "'is full?': #{stack.isFull?}"
  puts "'find 3': #{stack.search(3)}"
  puts "'peek': #{stack.peek}"
  stack.print

  (5..9).each do |num|
    stack.push(num)
  end

  puts 'ADD ITEMS 5 TO 9'
  puts "'size': #{stack.size}"
  puts "'is empty?': #{stack.isEmpty?}"
  puts "'is full?': #{stack.isFull?}"
  puts "'find 3': #{stack.search(3)}"
  puts "'peek': #{stack.peek}"
  stack.print

  puts "'pop': #{stack.pop}"
  puts "'size': #{stack.size}"
  puts "'is empty?': #{stack.isEmpty?}"
  puts "'is full?': #{stack.isFull?}"
  puts "'peek': #{stack.peek}"
  stack.print

  puts'GENERATE ERROR'
  stack.push(9)

  begin
    stack.push(10)
  rescue StandardError => e
    puts e.message
    stack.print
  end
end

